lib/nrf_modem_lib/trace_backends/uart/uart.c

```diff
@@ -1,142 +1,172 @@
-	const nrfx_uarte_config_t config = {
-		.skip_gpio_cfg = true,
-		.skip_psel_cfg = true,
-		.hal_cfg.hwfc = NRF_UARTE_HWFC_DISABLED,
-		.hal_cfg.parity = NRF_UARTE_PARITY_EXCLUDED,
-		.baudrate = NRF_UARTE_BAUDRATE_1000000,
-		.interrupt_priority = DT_IRQ(UART1_NL, priority),
-		.p_context = NULL,
-	};
+	struct uart_config uart_cfg;
 
 	if (trace_processed_cb == NULL) {
 		return -EFAULT;
 	}
 
 	trace_processed_callback = trace_processed_cb;
 
-	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(UART1_NL), PINCTRL_STATE_DEFAULT);
-	__ASSERT_NO_MSG(err == 0);
+	if (!device_is_ready(uart_dev)) {
+		printk("UART device not found!");
+		return -ENODEV;
+	}
+
+	err = uart_config_get(uart_dev, &uart_cfg);
+	if (err) {
+		LOG_ERR("UART_config_get failed: %d", err);
+		return err;
+	}
 
-	IRQ_CONNECT(DT_IRQN(UART1_NL), DT_IRQ(UART1_NL, priority), nrfx_isr,
-		    &nrfx_uarte_1_irq_handler, UNUSED_FLAGS);
+	if (uart_cfg.baudrate != TRACE_UART_BAUD_REQUIRED) {
+		LOG_ERR("Trace UART baudrate is %d but %d is required",
+			uart_cfg.baudrate, TRACE_UART_BAUD_REQUIRED);
+		return -ENOTSUP;
+	}
 
-	err = nrfx_uarte_init(&uarte_inst, &config, &uarte_callback);
-	if (err != NRFX_SUCCESS && err != NRFX_ERROR_INVALID_STATE) {
-		return -EBUSY;
+	err = uart_err_check(uart_dev);
+	if (err) {
+		LOG_ERR("UART check failed: %d", err);
+		return -EIO;
+	}
+
+	err = uart_callback_set(uart_dev, uart_callback, NULL);
+	if (err) {
+		LOG_ERR("Cannot set callback: %d", err);
+		return -EFAULT;
 	}
 
 	return 0;
 }
 
 int trace_backend_deinit(void)
 {
-	nrfx_uarte_uninit(&uarte_inst);
+	return 0;
+}
 
-#if CONFIG_PM_DEVICE
+/* Returns the number of bytes written, or negative error. */
+static int uart_send(const uint8_t *data, size_t len)
+{
 	int err;
 
-	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(UART1_NL), PINCTRL_STATE_SLEEP);
-	__ASSERT_NO_MSG(err == 0);
-#endif /* CONFIG_PM_DEVICE */
+	for (int i = 0; i < UART_TX_RETRIES; i++) {
+		err = uart_tx(uart_dev, data, len,
+			UART_TX_WAIT_TIME_MS * USEC_PER_MSEC);
+		if (err) {
+			LOG_ERR("uart error: %d", err);
+			return err;
+		}
+
+		err = k_sem_take(&tx_sem, K_FOREVER);
+		if (err) {
+			LOG_WRN("UARTE TX not available!");
+			return err;
+		}
+
+		if (tx_succeeded) {
+			return len;
+		}
+	}
 
 	return 0;
 }
 
 int trace_backend_write(const void *data, size_t len)
 {
-	nrfx_err_t err;
+	int err;
+	int ret;
 
 	/* Split RAM buffer into smaller chunks to be transferred using DMA. */
 	uint8_t *buf = (uint8_t *)data;
 	const size_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
 	size_t remaining_bytes = len;
 
+	err = k_sem_take(&tx_sem, K_FOREVER);
+	if (err) {
+		LOG_WRN("UARTE TX not available!");
+		return -EAGAIN;
+	}
+
 	while (remaining_bytes) {
 		size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
 		size_t idx = len - remaining_bytes;
 
-		if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
-			LOG_WRN("UARTE TX not available!");
+		ret = uart_send(&buf[idx], transfer_len);
```

Comment:
I'm not that familiar with how the modem tracing is implemented, but I think we should probably try to avoid corrupting data when sending it to the UART. If all traces can not be output, the modem will start dropping traces. The good thing about this is, that there's trace which indicates how much traces have been lost. If corruption/dropping of the traces happens at this level, we have no idea how much has been lost.

Because of this, I think `uart_send()` should always return the number of bytes it was able to send and that number should be given to `trace_processed_callback()`. Currently, when flow control activates, `uart_send()` returns zero although part of the data may have been sent. Also, `trace_processed_callback()` would not get called.